addons/account_tax_python/models/account_tax.py

```diff
@@ -62,50 +81,153 @@ def _ascending_process_taxes_batch(self, batch):
     def _eval_taxes_computation_prepare_product_fields(self, taxes_data):
         # EXTENDS 'account'
         field_names = super()._eval_taxes_computation_prepare_product_fields(taxes_data)
-        Product = self.env['product.product']
         for tax_data in taxes_data:
             if tax_data['amount_type'] == 'code':
-                tax = self.browse(tax_data['id'])
-                for formula in ((tax.python_applicable or '').strip(), (tax.python_compute or '').strip()):
-                    groups = REGEX_FORMULA_OBJECT.findall(formula)
-                    if groups:
-                        for group in groups:
-                            field_name = group[1]
-                            if field_name in Product:
-                                field_names.add(field_name)
-
+                field_names.update(tax_data['_product_fields'])
         return field_names
 
+    def _decode_formula(self, formula):
+        """ Decode the formula and extract relevant values from it.
+
+        :param formula: The value of the 'formula' field.
+        """
+        self.ensure_one()
+
+        if self.amount_type != 'code':
+            return {}
+
+        formula = (formula or '0.0').strip()
+        results = {
+            'js_formula': formula
+                .replace('and', '?')
+                .replace('or', ':')
+                .replace('None', 'undefined'),
+            'py_formula': formula,
+        }
+        product_fields = set()
+
+        groups = re.findall(r'((?:product\.)(?P<field>\w+))+', formula) or []
+        Product = self.env['product.product']
+        for group in groups:
+            field_name = group[1]
+            if field_name in Product:
+                product_fields.add(field_name)
+                results['py_formula'] = results['py_formula'].replace(f"product.{field_name}", f"product['{field_name}']")
+
+        results['product_fields'] = list(product_fields)
+        return results
+
+    @api.model
+    def _check_formula(self, tax_data):
+        """ Check the formula is passing the minimum check to ensure the compatibility between both evaluation
+        in python & javascript.
+
+        :param tax_data: The values returned by '_prepare_dict_for_taxes_computation'.
+        """
+        def startswith_number(formula, i):
```

Comment:
This inner function is particular. The name implied that we're checking if we're looking if we start with a number but it returns the number size to skip it after in decoding loop.
It's a bit confusing :sweat_smile: What about renaming the function name?
Like `get_number_size`?